ddtrace/settings/integration.py

```diff
@@ -1,62 +1,239 @@
 from copy import deepcopy
 
-from ..utils.attrdict import AttrDict
+from ..vendor import attr
 from ..utils.formats import asbool, get_env
 from .http import HttpConfig
 from .hooks import Hooks
 
 
-class IntegrationConfig(AttrDict):
+@attr.s
+class Setting(object):
+    """Setting represents a configuration setting which is defined by the
+    library and optionally overridden by the user.
     """
-    Integration specific configuration object.
+    @attr.s
+    class Undefined(object):
+        value = attr.ib(default=None)
 
-    This is what you will get when you do::
+    Undefined = Undefined()
+
+    @attr.s
+    class Defined(object):
+        value = attr.ib()
+
+    class SettingKeyError(Exception):
+        pass
+
+    _value = attr.ib(default=Undefined, validator=attr.validators.instance_of(Defined))
+    _default = attr.ib(init=False, default=Undefined)
+
+    def __attrs_post_init__(self):
+        self._default = self._value
+
+    def set(self, value):
+        assert value is not self.Undefined
+        assert not isinstance(value, Setting)
```

Comment:
It's never a good idea to do `assert` in the code; do a proper error checking and raise proper exceptions if you need to.

Reply:
Oh yeah, definitely. Just in there for my development purposes 😄